File: Part1._Enumeration_problem/Lesson4._Branch_and_Bound_technique/TravellingSalesman.pas
program TravellingSalesman;

const
    InputFile = 'TOURISM.INP';
    OutputFile = 'TOURISM.OUT';
    max = 20;
    maxC = 20 * 100 + 1;{+∞}
var
    C: array[1..max, 1..max] of Integer; {Ma trận chi phí}
    X, BestWay: array[1..max + 1] of Integer; {X để thử các khả năng, BestWay để ghi nhận nghiệm}
    T: array[1..max + 1] of Integer; {T[i] để lưu chi phí đi từ X[1] đến X[i]}
    Free: array[1..max] of Boolean; {Free để đánh dấu, Free[i]= True nếu chưa đi qua tp i}
    m, n: Integer;
    MinSpending: Integer; {Chi phí hành trình tối ưu}

procedure Enter;
var
    i, j, k: Integer;
    f: Text;
begin
    Assign(f, InputFile); Reset(f);
    ReadLn(f, n, m);
    for i := 1 to n do {Khởi tạo bảng chi phí ban đầu}
        for j := 1 to n do
            if i = j then C[i, j] := 0 else C[i, j] := maxC;
    for k := 1 to m do
        begin
            ReadLn(f, i, j, C[i, j]);
            C[j, i] := C[i, j]; {Chi phí như nhau trên 2 chiều}
        end;
    Close(f);
end;

procedure Init; {Khởi tạo}
begin
    FillChar(Free, n, True);
    Free[1] := False; {Các thành phố là chưa đi qua ngoại trừ thành phố 1}
    X[1] := 1; {Xuất phát từ thành phố 1}
    T[1] := 0; {Chi phí tại thành phố xuất phát là 0}
    MinSpending := maxC;
end;

procedure Try(i: Integer); {Thử các cách chọn xi}
var
    j: Integer;
begin
    for j := 2 to n do {Thử các thành phố từ 2 đến n}
        if Free[j] then {Nếu gặp thành phố chưa đi qua}
            begin
                X[i] := j; {Thử đi}
                T[i] := T[i - 1] + C[x[i - 1], j]; {Chi phí := Chi phí bước trước + chi phí đường đi trực tiếp}
                if T[i] < MinSpending then {Hiển nhiên nếu có điều này thì C[x[i - 1], j] < +∞ rồi}
                    if i < n then {Nếu chưa đến được x[n]}
                        begin
                            Free[j] := False; {Đánh dấu thành phố vừa thử}
                            Try(i + 1); {Tìm các khả năng chọn x[i+1]}
                            Free[j] := True; {Bỏ đánh dấu}
                        end
                    else
                        if T[n] + C[x[n], 1] < MinSpending then {Từ x[n] quay lại 1 vẫn tốn chi phí ít hơn trước}
                            begin {Cập nhật BestConfig}
                                BestWay := X;
                                MinSpending := T[n] + C[x[n], 1];
                            end;
            end;
end;

procedure PrintResult;
var
    i: Integer;
    f: Text;
begin
    Assign(f, OutputFile); Rewrite(f);
    if MinSpending = maxC then WriteLn(f, 'NO SOLUTION')
    else
        for i := 1 to n do Write(f, BestWay[i], '->');
    WriteLn(f, 1);
    WriteLn(f, 'Cost: ', MinSpending);
    Close(f);
end;

begin
    Enter;
    Init;
    Try(2);
    PrintResult;
end.

File: Part1._Enumeration_problem/Lesson4._Branch_and_Bound_technique/TravellingSalesman.c
#include <stdio.h>
#include <stdbool.h>

#define InputFile "TOURISM.INP"
#define OutputFile "TOURISM.OUT"
#define max 20

int C[max + 1][max + 1], Free[max + 1];
int X[max + 2], T[max + 2];
int *BestWay;
int m, n, MinSpending;
FILE *fi, *fo;

void Enter()
{
    fi = fopen("TOURISM.INP", "r");
    if (fi != NULL)
    {
        fscanf(fi, "%d %d\n", &n, &m);
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (i == j)
                {
                    C[i][j] = 0;
                }
                else
                {
                    C[i][j] = __INT_MAX__;
                }
            }
        }
        int i, j;
        for (int k = 1; k <= m; k++)
        {
            fscanf(fi, "%d %d %d\n", &i, &j, &C[i][j]);
            C[j][i] = C[i][j];
        }
        fclose(fi);
    }
}

void Init()
{
    for (int i = 0; i <= max; i++)
    {
        Free[i] = true;
    }
    Free[1] = false;
    X[1] = 1;
    T[1] = 0;
    MinSpending = __INT_MAX__;
}

void Try(int i)
{
    for (int j = 2; j <= n; j++)
    {
        if (Free[j])
        {
            X[i] = j;
            T[i] = T[i - 1] + C[X[i - 1]][j];
            if (T[i] < MinSpending)
            {
                if (i < n)
                {
                    Free[j] = false;
                    Try(i + 1);
                    Free[j] = true;
                }
                else if ((T[n] + C[X[n]][1]) < MinSpending)
                {
                    BestWay = X;
                    MinSpending = T[n] + C[X[n]][1];
                }
            }
        }
    }
}

void PrintResult()
{
    fo = fopen("TOURISM.OUT", "w");
    if (MinSpending == __INT_MAX__)
    {
        fprintf(fo, "NO SOLUTION\n");
    }
    else
    {
        for (int i = 1; i <= n; i++)
        {
            fprintf(fo, "%d->", BestWay[i]);
        }
        fprintf(fo, "1\n");
        fprintf(fo, "Cost: %d\n", MinSpending);
    }
    fclose(fo);
}

int main()
{
    Enter();
    Init();
    Try(2);
    PrintResult();
    return 0;
}